source/common/router/router.cc

```diff
@@ -2186,6 +2193,34 @@ void Filter::doRetry(bool can_send_early_data, bool can_use_http3, TimeoutRetry
     return;
   }
 
+  // Set router-set headers before host selection so request_headers_to_add can reference them.
+  if (include_attempt_count_in_request_) {
+    downstream_headers_->setEnvoyAttemptCount(attempt_count_);
+  }
+
+  if (include_timeout_retry_header_in_request_) {
+    downstream_headers_->setEnvoyIsTimeoutRetry(is_timeout_retry == TimeoutRetry::Yes ? "true"
+                                                                                      : "false");
+  }
+
+  // Update timeout headers for the retry attempt.
+  std::chrono::milliseconds elapsed_time = std::chrono::milliseconds::zero();
+  if (DateUtil::timePointValid(downstream_request_complete_time_)) {
+    Event::Dispatcher& dispatcher = callbacks_->dispatcher();
+    elapsed_time = std::chrono::duration_cast<std::chrono::milliseconds>(
+        dispatcher.timeSource().monotonicTime() - downstream_request_complete_time_);
+  }
+
+  FilterUtility::setTimeoutHeaders(elapsed_time.count(), timeout_, *route_entry_,
+                                   *downstream_headers_, !config_->suppress_envoy_headers_,
+                                   grpc_request_, hedging_params_.hedge_on_per_try_timeout_);
+
+  // Apply request_headers_to_add now that router-set headers are present.
+  const Formatter::Context header_transform_context(downstream_headers_, {}, {}, {}, {},
+                                                    &callbacks_->activeSpan());
+  route_entry_->finalizeRequestHeaders(*downstream_headers_, header_transform_context,
+                                       callbacks_->streamInfo(), !config_->suppress_envoy_headers_);
+
```

Comment:
When doing the retry, we may shouldn't call the `finalizeRequestHeaders` again because it may result in unexpected result. For example, if one of the header mutation is to `add` a new header, because the same header map object is used for retry, the same header may be added for multiple times which is unexpected result. (Although the users may want to map attempt count or timeout header to other headers, it's a corner case anyway. But adding a new normal header is a common case, we should keep the common case works properly first.)

So, IMO, we needn't make any code change to the doRetry and continueDoRetry